File: 백준/Gold/4179. 불！/불！.py
from collections import deque

n, m = map(int, input().split())
graph = [list(input()) for _ in range(n)]
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
q = deque()


def moveJ(x, y, c):
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < n and 0 <= ny < m:
            if graph[nx][ny] == ".":
                graph[nx][ny] = "J"
                q.append((nx, ny, c + 1))

    return


def moveF(x, y, c):
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < n and 0 <= ny < m:
            if graph[nx][ny] == "." or graph[nx][ny] == "J":
                graph[nx][ny] = "F"
                q.append((nx, ny, c + 1))

    return


def check(x, y, c):
    if x == 0 or x == n - 1 or y == 0 or y == m - 1:
        print(c)
        exit()


for i in range(n):
    for j in range(m):
        if graph[i][j] == "J":
            q.append((i, j, 1))
            break

for i in range(n):
    for j in range(m):
        if graph[i][j] == "F":
            q.append((i, j, 1))

while q:
    x, y, c = q.popleft()

    if graph[x][y] == "J":
        check(x, y, c)
        moveJ(x, y, c)
    else:
        moveF(x, y, c)

print("IMPOSSIBLE")

File: 백준/Gold/4179. 불！/불！.java
import java.util.*;
import java.io.*;

public class Main {
	static int n, m;
	static int[] dx = { 1, -1, 0, 0 };
	static int[] dy = { 0, 0, 1, -1 };
	static char[][] graph;
	static Queue q;

	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(bf.readLine());

		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());

		graph = new char[n][m];
		for (int i = 0; i < n; i++) {
			graph[i] = bf.readLine().toCharArray();
		}

		bfs();

		System.out.print("IMPOSSIBLE");
	}

	static void bfs() {
		q = new ArrayDeque<int[]>();

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (graph[i][j] == 'J') {
					q.add(new int[] { i, j, 1 });
				}
			}
		}

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (graph[i][j] == 'F') {
					q.add(new int[] { i, j, 1 });
				}
			}
		}

		while (!q.isEmpty()) {
			int[] out = (int[]) q.poll();
			if (graph[out[0]][out[1]] == 'J') {
				check(out[0], out[1], out[2]);
				moveJ(out[0], out[1], out[2]);
			} else {
				moveF(out[0], out[1], out[2]);
			}
		}
	}

	static void moveJ(int x, int y, int c) {
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (0 <= nx && nx < n && 0 <= ny && ny < m) {
				if (graph[nx][ny] == '.') {
					graph[nx][ny] = 'J';
					q.add(new int[] { nx, ny, c + 1 });
				}
			}
		}
	}

	static void moveF(int x, int y, int c) {
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (0 <= nx && nx < n && 0 <= ny && ny < m) {
				if (graph[nx][ny] == '.' || graph[nx][ny] == 'J') {
					graph[nx][ny] = 'F';
					q.add(new int[] { nx, ny, c + 1 });
				}
			}
		}
	}

	static void check(int x, int y, int c) {
		if (x == 0 || x == n - 1 || y == 0 || y == m - 1) {
			System.out.println(c);
			System.exit(0);
		}
	}
}
